validator/client/beacon-api/stream_blocks.go

```diff
@@ -0,0 +1,196 @@
+package beacon_api
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v3/beacon-chain/rpc/apimiddleware"
+	"github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+	"google.golang.org/grpc"
+)
+
+type abstractSignedBlockResponseJson struct {
+	Version             string          `json:"version" enum:"true"`
+	ExecutionOptimistic bool            `json:"execution_optimistic"`
+	Finalized           bool            `json:"finalized"`
+	Data                json.RawMessage `json:"data"`
+}
+
+type streamBlocksAltairClient struct {
+	grpc.ClientStream
+	ctx                 context.Context
+	beaconApiClient     beaconApiValidatorClient
+	streamBlocksRequest *ethpb.StreamBlocksRequest
+	prevBlockSlot       primitives.Slot
+	pingDelay           time.Duration
+}
+
+type headSignedBeaconBlockResult struct {
+	streamBlocksResponse *ethpb.StreamBlocksResponse
+	executionOptimistic  bool
+	slot                 primitives.Slot
+}
+
+func (c beaconApiValidatorClient) streamBlocks(ctx context.Context, in *ethpb.StreamBlocksRequest, pingDelay time.Duration) ethpb.BeaconNodeValidator_StreamBlocksAltairClient {
+	return &streamBlocksAltairClient{
+		ctx:                 ctx,
+		beaconApiClient:     c,
+		streamBlocksRequest: in,
+		pingDelay:           pingDelay,
+	}
+}
+
+func (c *streamBlocksAltairClient) Recv() (*ethpb.StreamBlocksResponse, error) {
+	result, err := c.beaconApiClient.getHeadSignedBeaconBlock(c.ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get latest signed block")
+	}
+
+	// We keep querying the beacon chain for the latest block until we receive a new slot
+	for (c.streamBlocksRequest.VerifiedOnly && result.executionOptimistic) || c.prevBlockSlot == result.slot {
```

Comment:
Why do you think there is a correlation between `VerifiedOnly` and optimistic execution?

Reply:
A bad assumption on my part... I assumed that "optimistic" blocks meant that they were not confirmed yet.

Is it simply a matter of `VerifiedOnly` being a `finalized` block, and `non-VerifiedOnly` being a head block?

Reply:
From https://github.com/ethereum/consensus-specs/blob/dev/sync/optimistic.md:
> In order to provide a syncing execution engine with a partial view of the head of the chain, it may be desirable for a consensus engine to import beacon blocks without verifying the execution payloads. This partial sync is called an optimistic sync.

I am not sure if we consider an optimistic block as verified, but nonetheless I don't think we should couple the two. I can imagine a scenario where the node is optimistically synced because the most recent block is optimistic, but the block coming from `streamBlocksRequest` is not optimistic. I am not sure how realistic this is, though.

Reply:
Makes sense. So what does `VerifiedOnly` refer to here? How would I get a non-verified block from the REST API, and what would it show up as in the response?

Reply:
`VerifiedOnly` means that the block has been received *and* fully processed (i.e. run through state transition, inserted into forkchoice etc.). `VerifiedOnly == false` means that the block has been received.

I don't think it's possible to get non-verified blocks from the Beacon API.

Reply:
Honestly I don't think we actually need this property, because we always set it to `true`...